hummingbot/connector/exchange/okx/okx_api_user_stream_data_source.py

```diff
@@ -0,0 +1,105 @@
+import asyncio
+from typing import TYPE_CHECKING, Any, Dict, Optional
+
+from hummingbot.connector.exchange.okx import okx_constants as CONSTANTS
+from hummingbot.connector.exchange.okx.okx_auth import OkxAuth
+from hummingbot.core.data_type.user_stream_tracker_data_source import UserStreamTrackerDataSource
+from hummingbot.core.web_assistant.connections.data_types import WSJSONRequest, WSPlainTextRequest, WSResponse
+from hummingbot.core.web_assistant.web_assistants_factory import WebAssistantsFactory
+from hummingbot.core.web_assistant.ws_assistant import WSAssistant
+from hummingbot.logger import HummingbotLogger
+
+if TYPE_CHECKING:
+    from hummingbot.connector.exchange.okx.okx_exchange import OkxExchange
+
+
+class OkxAPIUserStreamDataSource(UserStreamTrackerDataSource):
+
+    _logger: Optional[HummingbotLogger] = None
+
+    def __init__(
+            self,
+            auth: OkxAuth,
+            connector: 'OkxExchange',
+            api_factory: WebAssistantsFactory):
+        super().__init__()
+        self._auth: OkxAuth = auth
+        self._connector = connector
+        self._api_factory = api_factory
+
+    async def _connected_websocket_assistant(self) -> WSAssistant:
+        """
+        Creates an instance of WSAssistant connected to the exchange
+        """
+
+        ws: WSAssistant = await self._get_ws_assistant()
+        await ws.connect(
```

Comment:
This should be rate-limit protected on `CONSTANTS.WS_CONNECTION_LIMIT_ID`

Reply:
Good catch. I have applied the change.